### ISSUE TITLE
Error Sources/Security.php

### ISSUE THREAD

**Issue by devbeta** [SimpleMachines/SMF#4716]
Error : array_intersect(): Argument 2 is not an array

Sources/Security.php: 918
if (count(array_intersect($permission, $user_info['permissions'])) != 0)

Solution:

Sources/Security.php: 905
if (empty($user_info) || !isset($user_info['permissions']))

**Comment by frandominguezl** [comment]
How can we reproduce this error?

**Comment by devbeta** [comment]
In the error log messages from the guest, possibly through SSI.

**Comment by albertlast** [comment]
could you test if you issue get fixed when you place this check here:
https://github.com/SimpleMachines/SMF2.1/blob/dd179b23c765f51a490c9464e4db0cd06aa86137/Sources/Load.php#L1950
```
if (!isset($user_info['permissions']))
    $user_info['permissions'] = array();
```

like d3vcho mention we need to know what the root cause is.

**Comment by MissAllSunday** [comment]
This has been reported before, can't remember which feature it was but it added it a check on permissions way too early in the flow.  Since this is about permissions, I'm afraid we cannot simple declare the array as empty as it will mean everyone will everyone will have the permission to do whatever that check is performing. We need to figure it out what exactly is doing and see if we can defer it to after the permissions has been loaded which shouldn't be an issue since those gets loaded pretty early.

**Comment by frandominguezl** [comment]
We still don't have a way to reproduce this error.

**Comment by wintstar** [comment]
I had already had this error message. But I accidentally deleted the error log. I had disabled the feature that guests can browse the forum. But now I can't specify 100% that the error message comes from this.

**Comment by Archer70** [comment]
This is the same error that was appearing in #4801. It happens when `$user_info` is overwritten with loadMemberData(), because that function doesn't load everything that's normally in $user_info.

"Key" differences are:
- is_guest
- is_admin
- buddies
- permissions

In the case of the alerts bug, it was simple to fix by reverting `$user_info` back to its previous state once the alerts function was done doing its magic. The logic for reverting it was already in place, but had a bug causing it to revert to `null`.

If anyone comes across this error again, please provide a stack trace so we can track down where `$user_info` is being screwed up. Thanks :)

**Comment by jdarwood007** [comment]
@MissAllSunday A empty array should not be granting access.  If that is so, we have a bigger issue at hand.  In this case, the error as generated should be causing a fail safe and denying permissions even if they have access.

The reason declaring a empty array isn't what we want to do to fix this is because we need to know why.   The process that gets things loaded up should have already loaded $user_info['permissions'] to at least something even if not all permissions have been loaded up.

**Comment by Archer70** [comment]
> The process that gets things loaded up should have already loaded $user_info['permissions'] to at least something even if not all permissions have been loaded up.

According to my tests in the alerts case, permissions were previously loaded, but this bit in the fetch_alerts() function was overwriting $user_info with a new copy that didn't contain permissions.

```php
// are we someone else?
if (empty($user_info) || $user_info['id'] !== $memID)
{
	$user_old = !empty($user_info) ? $user_info : null;
	if (empty($user_profile[$memID]))
		loadMemberData($memID, false, 'profile');
	$user_info = $user_profile[$memID];
}
```

My guess is something similar is happening in this case, but we need to know where the error is occurring and get a stack trace if at all possible.

**Comment by jdarwood007** [comment]
That would only affect 2.1. This bug has existed for many years going back to 1.1 or further (can't remember).

Highly odd we are switching profiles like that.

**Comment by sbulen** [comment]
Upon a 2nd read, I like the solution proposed in the initial report above.  
It's not overwriting any data.  It's not changing permissions at all.  
It simply returns false if permissions haven't been set yet.  

I think it may have gotten short shrift because it was hard to see the overall flow given that one line...  

Although it would be nice to know when/how this is being called, I do think it is safe to return false for allowedTo() prior to any permissions being set.  

Basically, early in allowedTo(), change
```
	// You're never allowed to do something if your data hasn't been loaded yet!
	if (empty($user_info))
		return false;
```
To:
```
	// You're never allowed to do something if your data hasn't been loaded yet!
	if (empty($user_info) || !isset($user_info['permissions']))
		return false;
```

You almost certainly want it to return false when permissions haven't yet been set, correct?  Like the comment says?

**Comment by Sorck** [comment]
@sbulen 
Would a fatal_lang_error not be a better choice? Rather than the user not being allowed to do this current action, there is actually an error in SMF applicatino state - we should not have any calls to allowedTo before the user profile is loaded. This would also make it easier to catch where any calls were coming from.

The only issue is if the call to allowedTo is in the template. In this case we may just have to log the error and return false from allowedTo as I don't know the consequences of a fatal_lang_error occurring after template output has begun. Probably depends on if the output buffer is being used?

@jdarwood007 
The switching of profiles in ```fetch_alerts()``` doesn't seem to have any purpose.

It breaks any error logging within fetch_alerts (attributing it to $memID rather than the *real* $user_info['id']) and also breaks the permissions checking capability of the integration hook.

The SMF function calls within fetch_alerts are:
* timeformat
* loadMemberData
* loadMemberContext
* loadLanguage
* integrate_fetch_alerts

All of these functions user $user_info in a way that may not be correctly loaded in this configuration, yet none of them appear to benefit from this profile switch.

**Comment by sbulen** [comment]
@Sorck - this function is already designed to simply return false if called too early without permissions being properly set yet.   The edge issue here is that $user_info is set but $user_info['permissions'] is not.

**Comment by Sorck** [comment]
# A backtrace for the error

I have backtrace created when a user signs up and you hover the alert.

## Error 
    Type of error: Critical
    Error message: allowedTo() called when $user_info was partially loaded.
    URL of page causing the error: {$scripturl}?action=profile;area=alerts_popup;u=1

## Backtrace information
    #0: allowedTo()
    Called from {$sourcedir}/Load.php on line 1432
    #1: loadMemberContext()
    Called from {$sourcedir}/Profile-View.php on line 266
    #2: fetch_alerts()
    Called from {$sourcedir}/Profile.php on line 843
    #3: alerts_popup()
    Called from {$sourcedir}/Profile.php on line 734
    #4: ModifyProfile()
    Called from {$sourcedir}/../index.php on line 150

Please note this uses a patch I have applied that raises a critical error when allowedTo is illegally called.

The problem occurrs in loadMemberContext https://github.com/SimpleMachines/SMF2.1/blob/a0e410fc2d260b26078aa222de4f87a8f003b5d6/Sources/Load.php#L1449 and another possible issue would be at https://github.com/SimpleMachines/SMF2.1/blob/a0e410fc2d260b26078aa222de4f87a8f003b5d6/Sources/Load.php#L1427

# Finding the cause

## Where is $user_info overwritten?
As mentioned earlier, it is overwritten in Profile-View.php fetch_alerts() https://github.com/SimpleMachines/SMF2.1/blob/a0e410fc2d260b26078aa222de4f87a8f003b5d6/Sources/Profile-View.php#L238

## Surely it shouldn't be overwritten?
It certainly shouldn't. However, PHP has no static typing - the error is caused by $memID having been passed to fetch_alerts() as a string by alerts_popup(), which in turn got it from ModifyProfile(). However, $user_info['id'] is an integer. Therefore the equality check fails.

## A bigger issue?
This is actually a maze of type issues that could be impacting SMF as a lot of people would, quite rightly, assume id was an integer, though some places also assume it's a string.

# Fixes
There are at least a couple of solutions:
1. update the loadMemberData and everything that uses an ID to enforce them being an integer. This would take a while but be good for type safety in the long run.
2. coerce the type (int) before calling fetch_alerts() (a strict type check could be added in fetch_alerts to insist that that argument is an integer, or we could just silently coerce, or do a bit of both - log the error as a type error but continue with a coercision)

Personally I'd like to see the 'fuzzy' type checks in options 2 - i.e. log a critical error (invalid argument type) but silently continue with a coercsion. The same can be done in allowedTo - log it as a critical error (call to allowedTo before permissions loaded) but continue assuming the user has no permissions.

The first option would also be good, but it could open up a huge long line of things to correct.

**Comment by jdarwood007** [comment]
@Sorck Not sure how $row['id_member'] could return anything other than numerical, but otherwise looks good. See #4834 as it resolves the alerts overriding the user_info.  Does not solve the root cause to permissions not existing, but resolves the current case you mentioned.

**Comment by Sorck** [comment]
@jdarwood007 I tested $row['id_member'] with a gettype() in loadMemberData() so it does seem to be a string. [Some quick reading](https://stackoverflow.com/questions/5323146/mysql-integer-field-is-returned-as-string-in-php) suggests this may be MySQL specific.

Commit Sorck/SMF2.1@4ae148e571b1d95b2a2d97af654e21014c756d83 actually stops the overwrite of profile from happening in the regular use case as the comparison of $memID and $user_info['id'] is equal if the patch is applied, so the profile never gets swapped out under normal use.

I think making things integers makes sense alongside removing the profile swap if possible.

More generally the lack of the permissions array shouldn't be a case that happens unless something overwrites $user_info or a ridiculously early call to allowedTo() is made - perhaps a more descriptive error is all we need? E.g.
```php
// in allowedTo()
if (empty($user_info))
    fatal_lang_error('user_not_loaded', 'critical', array(), 500);
if (!isset($user_info['permissions']))
    fatal_lang_error('user_permissions_not_loaded', 'critical', array(), 500);
if (!isset($user_info['is_admin']))
    fatal_lang_error('user_admin_not_loaded', 'critical', array(), 500);
// language files
$txt['user_not_loaded'] = 'allowedTo() or isAllowedTo() was called before $user_info was loaded, or the $user_info array has been deleted.';
$txt['user_permissions_not_loaded'] = 'allowedTo() or isAllowedTo() was called before the user permissions were loaded, or the $user_info array has been corrupted.';
$txt['user_admin_not_loaded'] = 'allowedTo() or isAllowedTo() was called before the user was determined to be a regular user of admin, or the $user_info array has been corrupted.';
```

**Comment by jdarwood007** [comment]
Will need some more testing, but good reference point.  Indeed that seems like mysql extension in php is just being stupid here.

**Comment by albertlast** [comment]
@jdarwood007 is the issue still open?

**Comment by jdarwood007** [comment]
I haven't got the error.  I still have concerns over the security of the fetch_alerts function.  But it seems that issue isn't a high priority as nobody can explain why we have to overload $user_info for that function.

**Comment by jdarwood007** [comment]
This is also a long standing bug that has existed since 2.0 and before.  So to say this is the fix is hard.  We can close it, but it may return if wider testing finds it again.
